src/binding/c/request_api.txt

```diff
@@ -0,0 +1,617 @@
+
+MPIX_Grequest_class_allocate:
+    greq_class: GREQUEST_CLASS, [generalized request class]
+    extra_state: EXTRA_STATE, [extra state]
+    request: REQUEST, direction=out, [generalized request]
+    .desc: Create and return a user-defined extended request based on a generalized request class
+
+MPI_Request_free:
+    .desc: Frees a communication request object
+    .seealso: MPI_Isend, MPI_Irecv, MPI_Issend, MPI_Ibsend, MPI_Irsend, MPI_Recv_init, MPI_Send_init, MPI_Ssend_init, MPI_Rsend_init, MPI_Wait, MPI_Test, MPI_Waitall, MPI_Waitany, MPI_Waitsome, MPI_Testall, MPI_Testany, MPI_Testsome
+/*
+    Notes:
+
+    This routine is normally used to free inactive persistent requests created with
+    either 'MPI_Recv_init' or 'MPI_Send_init' and friends.  It `is` also
+    permissible to free an active request.  However, once freed, the request can no
+    longer be used in a wait or test routine (e.g., 'MPI_Wait') to determine
+    completion.
+
+    This routine may also be used to free a non-persistent requests such as those
+    created with 'MPI_Irecv' or 'MPI_Isend' and friends.  Like active persistent
+    requests, once freed, the request can no longer be used with test/wait routines
+    to determine completion.
+*/
+
+MPI_Request_get_status:
+    .desc: Nondestructive test for the completion of a Request
+/*
+    Notes:
+    Unlike 'MPI_Test', 'MPI_Request_get_status' does not deallocate or
+    deactivate the request.  A call to one of the test/wait routines or
+    'MPI_Request_free' should be made to release the request object.
+*/
+{ -- early_return --
+    if (request == MPI_REQUEST_NULL) {
+        *flag = 1;
+        MPIR_Status_set_empty(status);
+        if (status != MPI_STATUS_IGNORE) {
+            status->MPI_ERROR = MPI_SUCCESS;
+        }
+        goto fn_exit;
+    }
+}
+
+MPI_Start:
+    .desc: Initiates a communication with a persistent request handle
+{
+    mpi_errno = MPID_Startall(1, &request_ptr);
+    if (mpi_errno != MPI_SUCCESS)
+        goto fn_fail;
+}
+
+MPI_Startall:
+    .desc: Starts a collection of persistent requests
+    .impl: mpid
+/*
+    Notes:
+    Unlike 'MPI_Waitall', 'MPI_Startall' does not provide a mechanism for
+    returning multiple errors nor pinpointing the request(s) involved.
+    Furthermore, the behavior of 'MPI_Startall' after an error occurs is not
+    defined by the MPI standard.  If well-defined error reporting and behavior
+    are required, multiple calls to 'MPI_Start' should be used instead.
+*/
+
+MPI_Status_set_cancelled:
+    .desc: Sets the cancelled state associated with a request
+{
+    MPIR_STATUS_SET_CANCEL_BIT(*status, flag ? TRUE : FALSE);
+}
+
+MPI_Test_cancelled:
+    .desc: Tests to see if a request was cancelled
+    .docnotes: SignalSafe
+{
+    *flag = MPIR_STATUS_GET_CANCEL_BIT(*status);
+}
+
+MPI_Test:
+    .desc: Test for the completion of a request
+    .docnotes: waitstatus
+{
+    mpi_errno = MPIR_Test(request, flag, status);
```

Comment:
> In test, `MPIR_ERRTEST_COUNT(count, mpi_errno);` is lost,

Did you mean in e.g. `MPI_Test{any,all,some}`? There is no count in `MPI_Test`. Anyway, this should be fixed with https://github.com/pmodels/mpich/pull/4967#discussion_r548780055

> and no `MPI_ERR_REQUEST`.

This is fixed with the "inout" handle python commit.

> maybe better to add back the note for temporary comm_ptr, since it is not straightforward until see fn_fail.

Added a commit to update python script to generate this comment.